### ISSUE TITLE
Static field offsets encoded in DWARF debug info are incorrect when isolates are enabled

### ISSUE THREAD

**Issue by adinn** [oracle/graal#4224]
**Issue description**
Printing of static fields in gdb is no longer working correctly when debugging a native image created using -H:+SpawnIsolates. The offset of the static field from the heap register that is encoded into the debug records appears to be less than the correct value by 0x100000.

Note that this functionality was working previously but has not been regularly tested because of a problem in gdb 9 & 10. The problem can be manifested when using gdb 11.

Note also that the offset and subsequent printing are correct if the native image is compiled with -H:-SpawnIsolates. 

**Steps to reproduce the issue**
Ensure you have gdb 11 in your path
Clone the latest graal repo
Run the following command
```
$ (export GDB_CAN_PRINT=True ; mx debuginfotest )
  . . .
(gdb) print 'java.math.BigDecimal'::BIG_TEN_POWERS_TABLE
$3 = (_z_.java.math.BigInteger[] *) 0x0

(gdb) print 'java.math.BigDecimal'::BIG_TEN_POWERS_TABLE->data[3]->mag->data[0]
Traceback (most recent call last):
  File "/home/adinn/redhat/openjdk/graal/graal/substratevm/mx.substratevm/testhello.py", line 655, in <module>
    test()
  File "/home/adinn/redhat/openjdk/graal/graal/substratevm/mx.substratevm/testhello.py", line 278, in test
    exec_string = execute("print 'java.math.BigDecimal'::BIG_TEN_POWERS_TABLE->data[3]->mag->data[0]")
  File "/home/adinn/redhat/openjdk/graal/graal/substratevm/mx.substratevm/testhello.py", line 115, in execute
    return gdb.execute(command, to_string=True)
gdb.error: Location address is not set.
(gdb) 
```
At this point you need to type q to quit the failing scripted gdb session.

**Describe GraalVM and your environment:**
 - GraalVM version: latest head
 - JDK major version: 11
 - OS: Linux
 - Architecture: x86/aarch64

**Further info**

At the quit prompt above it is possible to identify the offset being used for field
`java.math.BigDecimal::BIG_TEN_POWERS_TABLE` as follows
```
(gdb) p &'java.math.BigDecimal'::BIG_TEN_POWERS_TABLE
$4 = (_z_.java.math.BigInteger[] **) 0x7ffff7758ce8
(gdb) p &'java.math.BigDecimal'::BIG_TEN_POWERS_TABLE - $r14
$5 = (_z_.java.math.BigInteger[] **) 0xfffc80003df58ce8
(gdb) p/x 0x7ffff7758ce8L - $r14
$6 = 0x458ce8
```
The correct offset is actually greater by `0x100000`. The following commands
- show the compressed object rererence at the adjusted offset
- print it as a `BigInteger[] `
- show the characters in the `String` located in the hub name field
- print element 0 of the mag field in the 3rd element of the BigInteger[] (10^3)
```
(gdb) p/x $r14 + 0x558ce8
$7 = 0x7ffff7858ce8
(gdb) x/gx 0x7ffff7858ce8
0x7ffff7858ce8:	0x000000000058e910
(gdb) p *('_z_.java.math.BigInteger[]' *)0x000000000058e910
$8 = {
  <java.math.BigInteger[]> = {
    <java.lang.Object> = {
      <_objhdr> = {
        hub = 0x437c42,
        idHash = 1081058246
      }, <No data fields>}, 
    members of java.math.BigInteger[]:
    len = 19,
    data = 0x7ffff788e920
  }, <No data fields>}
(gdb) p *(('_z_.java.math.BigInteger[]' *)0x000000000058e910)->hub->name->value
$9 = {
  <byte []> = {
    <java.lang.Object> = {
      <_objhdr> = {
        hub = 0x42fe48,
        idHash = 1811160057
      }, <No data fields>}, 
    members of byte []:
    len = 23,
    data = 0x7ffff74c9e00 "[Ljava.math.BigInteger;"
  }, <No data fields>}
(gdb) p (('_z_.java.math.BigInteger[]' *)0x000000000058e910)->data[3]->mag->data[0]
$10 = 1000
```
The same disparity in offset occurs for all other static fields.

**Problem Diagnosis**
The problem is in the reporting of the base position for the arrays holding primitive and object static values as can be seen by stepping through the relevant code in class `NativeImageDebugInfoProvider`. Static field offsets are computed in two steps. When debug info generation starts `NativeImageDebugInfoProvider` computes the heap [base offsets](https://github.com/oracle/graal/blob/cdcb47194d475c2aaa0f78a1e3f5861a332b9b08/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java#L122) for the `staticPrimitiveFields` and `staticReferenceFields` arrays. When reporting a specific static field it adds the [field location value](https://github.com/oracle/graal/blob/cdcb47194d475c2aaa0f78a1e3f5861a332b9b08/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java#L600) to the appropriate base offset to obtain the offset for the field.

The second step is using the correct value. The first step is returning values that are out by 0x100000. The required offset can be found in several ways. I worked it out by debugging a program that assigned a static field (an application shutdown hook) and noting the disparity between the notified offset and the actual offset used in the compiled code. However,it is most easily observed by by enabling verbose trace (with option `-H:Log=DebugCodeInfo:3`) and searching the log output for `Data:` trace log records containing the strings `staticObjectFields` and `staticReferenceFields`.
```
$ grep -ir staticobjectfields out.log | grep Data:
      Data: address 0x41e598 size 0x190 type java.lang.Object[] partition writableReference{1746584}@3029368 provenance [Ljava.lang.Object; -> staticObjectFields 
$ grep -ir staticprimitivefields out.log | grep Data:
      Data: address 0x3cff10 size 0xc8 type byte[] partition writablePrimitive{588152}@2441216 provenance [B -> staticPrimitiveFields 
```
These trace statements dump details of `ObjectInfo` instances streamed via NativeImageHeap::getObjects() i.e. in the above cases they are using the same instances as those used in the code cited earlier to retrieve offsets. The address value is computed by calling `ObjectInfo.getAddress()`. So, with `-H:+SpawnIsolates` the address and offset differ by 0x100000. When the same info is dumped for an image built with `-H:-SpawnIsolates` the address and offset values are the same.

**Possible Solution**
Clearly a solution to this problem is for the debuginfo code to use `getAddress()` when computing the heap base offsets. Although this might be correct for the code as it stands at present it is not clear that this will is actually be the correct way to calculate the offset that needs to be added to the heap base register (or, for the no isolates case, to the address identified by symbol `_svm_heap_begin`). The names of the relevant methods suggest otherwise. It would be helpful to have some clarification of what getOffset and getAddress are intended to report and why they might differ before proceeding to fix this.

**Comment by olpaw** [comment]
> Clearly a solution to this problem is for the debuginfo code to use getAddress() when computing the heap base offsets.
 
`getAddress()` is the right method.

`com.oracle.svm.hosted.image.NativeImageHeap.ObjectInfo#getOffset` only gives you the offset within the ImageHeap. To get the actual offset from the base-register you use `com.oracle.svm.hosted.image.NativeImageHeap.ObjectInfo#getAddress()` which factors in the actual position of the image heap in the address space (determined by `HeapParameters.Options#AlignedHeapChunkSize` & others).

**Comment by zakkak** [comment]
Fixed in #4226
